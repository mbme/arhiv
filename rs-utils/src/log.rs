use log::LevelFilter;
use tracing_subscriber::{
    fmt::{self, format::FmtSpan, time::ChronoLocal},
    layer::SubscriberExt,
    util::SubscriberInitExt,
};

pub use log::{debug, error, info, trace, warn, Level};

const DEFAULT_LOG_LEVELS: &str =
    "hyper=info,h2=info,rustls=info,mdns_sd=info,axum::rejection=trace";

#[cfg(target_os = "android")]
pub fn setup_android_logger(package: &str) {
    tracing_subscriber::registry()
        .with(tracing_subscriber::EnvFilter::new(format!(
            "debug,{DEFAULT_LOG_LEVELS}"
        )))
        .with(tracing_android::layer(package).expect("failed to build android tracing subscriber"))
        .init();
}

#[cfg(not(target_os = "android"))]
pub fn setup_android_logger(_package: &str) {
    unreachable!()
}

fn setup_logger_with_level(log_level: LevelFilter) {
    tracing_subscriber::registry()
        .with(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| format!("{log_level},{DEFAULT_LOG_LEVELS}").into()),
        )
        .with(
            fmt::Layer::new()
                .compact()
                .with_timer(ChronoLocal::new("[%Y-%m-%d][%H:%M:%S]".to_string()))
                .with_span_events(FmtSpan::CLOSE),
        )
        .init();
}

pub fn setup_error_logger() {
    setup_logger_with_level(LevelFilter::Error);
}

pub fn setup_warn_logger() {
    setup_logger_with_level(LevelFilter::Warn);
}

pub fn setup_logger() {
    setup_logger_with_level(LevelFilter::Info);
}

pub fn setup_debug_logger() {
    setup_logger_with_level(LevelFilter::Debug);
}

pub fn setup_trace_logger() {
    setup_logger_with_level(LevelFilter::Trace);
}
