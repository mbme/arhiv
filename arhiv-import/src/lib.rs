use anyhow::*;
use dialoguer::{theme::ColorfulTheme, Confirm};

use arhiv_core::{entities::Id, Arhiv};
use rs_utils::log;

use crate::{
    book_yakaboo::YakabooBookImporter,
    utils::{scrape, Importer},
};

mod book_yakaboo;
mod utils;

pub struct ArhivImport {
    arhiv: Arhiv,
    confirm: bool,
}

impl ArhivImport {
    pub fn new(arhiv: Arhiv) -> Self {
        ArhivImport {
            arhiv,
            confirm: true,
        }
    }

    pub fn confirm(&mut self, confirm: bool) -> &mut Self {
        self.confirm = confirm;

        self
    }

    fn import_confirmed(&self) -> Result<bool> {
        if !self.confirm {
            return Ok(true);
        }

        Confirm::with_theme(&ColorfulTheme::default())
            .with_prompt("Do you really want to continue?")
            .default(true)
            .interact()
            .context("failed to ask confirmation")
    }

    pub async fn import(self, url: &str) -> Result<Option<Id>> {
        log::info!("Importing url {}", url);

        let importers: Vec<Box<dyn Importer>> = vec![
            Box::new(YakabooBookImporter), //
        ];

        for importer in importers {
            if !importer.can_import(url) {
                continue;
            }

            log::info!("Found importer {}", importer.get_name());

            log::info!("Scraping data ...");
            let data = scrape(url).context("scrape failed")?;
            log::info!("Scraped data:\n{}", &data);

            if !self.import_confirmed()? {
                return Ok(None);
            }

            let document = importer
                .import(&data, &self.arhiv)
                .await
                .context("importer failed")?;

            log::info!("Imported {}", document);

            return Ok(Some(document.id));
        }

        bail!("don't know how to import document from url '{}'", url)
    }
}
