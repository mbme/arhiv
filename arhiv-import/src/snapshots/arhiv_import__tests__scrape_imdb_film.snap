---
source: arhiv-import/src/lib.rs
expression: result

---
{
  "cast": "Keanu Reeves, Laurence Fishburne, Carrie-Anne Moss",
  "countries_of_origin": "United States, Australia",
  "cover": "[cover_attachment_id]",
  "creators": "Lana Wachowski, Lilly Wachowski",
  "description": "Thomas A. Anderson is a man living two lives. By day he is an average computer programmer and by night a hacker known as Neo. Neo has always questioned his reality, but the truth is far beyond his imagination. Neo finds himself targeted by the police when he is contacted by Morpheus, a legendary computer hacker branded a terrorist by the government. As a rebel against the machines, Neo must confront the agents: super-powerful computer programs devoted to stopping Neo and the entire human rebellion.",
  "duration": "2h 16m",
  "is_series": false,
  "original_language": "English",
  "release_date": "1999",
  "title": "The Matrix"
}
